### ISSUE TITLE
Can't undo "always open this site in..." (i.e. can't revert to opening a url in default container)

### ISSUE THREAD

**Issue by aminomancer** [mozilla/multi-account-containers#1996]
- Multi-Account Containers Version: 7.3.0
- Operating System + Version: Windows 10, 20H2
- Firefox Version: 88.0a1, nightly/mozilla-central
- Other installed Add-ons + Version + Enabled/Disabled-Status: dozens of extensions but this bug exists whether they're disabled or not

### STR
- Go to a website, e.g. www.github.com
- Use the page action button to choose a container for the selected tab's url.
- Open a new tab, navigate to the url.
- Extension prompt asks if you want to open the url in the current container or in the container you selected with the page action.
- It also asks if you want the extension to remember your decision for this site.
- Click "remember my decision for this site" and then choose to open in the current container.
- The url opens in the default container.
- Now close the tab, open a new tab, navigate to the url again.

### Actual behavior
- The extension prompt comes up again, asking if you want to open the url in the current (default) container or in the container you previously chose. It does not remember that you chose to open the url in the default container.
- Now every time you open www.github.com, an annoying prompt comes up and forces interaction before you can proceed to the site.
- The only way I could find to eliminate this prompt is to delete the key in storage, `siteContainerMap@@_github.com:"{"userContextId":"1","neverAsk":false,"identityMacAddonUUID":"c2836c9e-0bde-4069-9a0b-1e3a0ca3a88d"}"`.
- But I imagine that is way beyond the lengths the average user is likely to go to get rid of this. Probably one would try to uninstall and reinstall the extension, only to find that extension storage survives uninstallation. If I didn't already know how to clear extension storage I guess I would have just uninstalled the extension and never used it again, unfortunately.

### Expected behavior
The extension should be able to remember the user's choice to open a url in the *default* container, just like it can remember the user's choice to open a url in a unique container.

**Comment by aminomancer** [comment]
Also this is unrelated, but might as well mention the page action popup's icons are broken for me. It looks like the style attribute `style="enable-background:new 0 0 32 32;"` in /img/usercontext.svg is causing the fill filter to color the entire image area, making all the icons square. I was able to fix it like this `svg:root { background: none !important; }` although maybe that messes up some icons I haven't seen yet.

**Comment by ameaninglessname** [comment]
![image](https://user-images.githubusercontent.com/20263103/112023264-0ca90900-8b6e-11eb-861a-a781bfc13927.png)
Have you tried this right-click context menu?

**Comment by aminomancer** [comment]
For whatever reason the context menu doesn't appear at all for me, haven't seen it in probably over a year

**Comment by leplatrem** [comment]
> Have you tried this right-click context menu?

Thanks! That worked for me. 

Once the container has been selected:
- right click on the page
- bottom of context menu: Firefox Multicontainers
- Uncheck *Always open in this container*

**Comment by grahamperrin** [comment]
> context menu doesn't appear

1. At an affected page
2. use the context menu against a 'dead' area of the page. 

For example, near the top left of a GitHub issue page: 

![image](https://user-images.githubusercontent.com/192271/115880447-a6e8bf00-a442-11eb-9cd6-6c9d7f7e3425.png)

**Comment by grahamperrin** [comment]
> … remember the user's choice to open a url in the _default_ container …

If by _default_ you mean _non-contained_, then this #1996 is a duplicate of #1967. Please consider closing this issue, for open (non-duplicated) issues to be more easily managed.

**Comment by aminomancer** [comment]
it doesn't appear whether the default container is set to null or set to an actual container

**Comment by grahamperrin** [comment]
> it doesn't appear

What doesn't appear?

(Sorry … working on dozens of issues, with so many duplicates … need things to be non-ambiguous. Thanks.)

**Comment by grahamperrin** [comment]
> it doesn't appear …

Can you provide a screen recording? It's difficult for me visualise the gap in this part of your UX. Thanks.

**Comment by grahamperrin** [comment]
In the wiki: 

* [How to remove a "Awlays open in…" site from a Container · mozilla/multi-account-containers Wiki](https://github.com/mozilla/multi-account-containers/wiki/How-to-remove-a-%22Awlays-open-in%E2%80%A6%22-site-from-a-Container)

**Comment by aminomancer** [comment]
I mean the context menu for the extension. Just doesn't show up on any page. Haven't seen it in like a year or more. It's not hidden or something, it's literally not in the DOM. Just menus/menuitems for all my other extensions, plus the internal menuitems. As for the toolbar button popup, it's good that there's at least one way to forget a setting, but I don't use the toolbar button, I use the page action button and the menupopup that appears when you click & hold the new tab button.

Seems like this functionality should be duplicated in at least the page action panel, since it's contextually relevant to the selected tab's container. Like, if I use the page action button to tell the extension to always open the current tab's URL in a container, I should be able to use the page action button to reverse the same action.

Or at the very least, when navigating to a URL that's been associated with a container, if it loads the intermediate landing page that asks if I want to load the page in the current container or in the associated container, I should then be able to make it forget the association altogether rather than merely switching the association from one container to another.

Like, if you're able to change from one container association to another, you should also be able to switch from a container to no container. I understand that internally, "no container" is not semantically the same as a container, but from the user's point of view, the non-container seems like it's just the main container.

I get that the context menu solves that problem for some people, but the fact that some functionality exists in one place isn't a good justification for not including that functionality anywhere else. Even if the context menu works, it still only lets you forget associations for whatever URI is active in the selected browser at the time the context menu is called. So if you want to forget associations for a lot of sites quickly, you need to use the toolbar button. This seems weird because the extension already has an options page, (that is, in about:addons) but it has even less functionality than the toolbar button popup.

And there's a page action panel too but all it can do is make an association, not undo it. That just seems silly to me. It makes sense for the toolbar button popup to have the most utility but the page action button should at least be able to do the same things the context menu can do, if not provide some portal to the "manage container" interface. That way since the context menu never loads for me, I can use something else without needing to permanently add a widget to the customizable navbar

**Comment by aminomancer** [comment]
<img width="636" alt="ff-multi_account_containers_contextmenu" src="https://user-images.githubusercontent.com/33384265/115959306-76476a80-a4c0-11eb-95bd-6174db76e8be.png">

here's a screenshot of the context menu for this page, which is currently associated with the "Work" container

**Comment by aminomancer** [comment]
<img width="605" alt="ff-multi_account_containers_contextmenu_DOM" src="https://user-images.githubusercontent.com/33384265/115959427-071e4600-a4c1-11eb-99ff-17a02afccf50.png">

a screenshot of the DOM

**Comment by grahamperrin** [comment]
Hmm, I recall adding a long comment, it disappeared. My bad, maybe I accidentally closed the tab. 

tl;dr this _might_ be an example of a type of page with which the context menu item will not appear: 

* https://forums.virtualbox.org/viewtopic.php?f=1&t=90991#p436905

@aminomancer can you provide a list of enabled extensions? 

[Add-on List-o-matic 9000](https://addons.mozilla.org/addon/add-on-list-o-matic-9000/)

* the names alone, please
* nothing else
* alphabetical.

In the meantime I'll redraft my missing comment. tl;dr igoring the example URL above, I do occasionally find the context menu without a menu item for Multi-Account Containers, however most such incidents are transient, I attribute them to my **extraordinarily** heavyweight use of extensions (not an issue with the Multi-Account Containers extension).

**Comment by aminomancer** [comment]
Well I personally don't see the context menu on *any* site, I don't think it's even registering a context menu item. I haven't seen it on a single site in over a year, and I have presumably visited several thousands of sites in that span of time. I use firefox like 6 hours a day at least, I definitely would have noticed if it was appearing on some sites and not others. When I first noticed that it was gone, I thought maybe I accidentally hid it with some CSS rule or messed it up with a script. But I tested it without all that stuff and it was still gone. It does show up on a fresh profile, but that's not really a solution. Maybe the extension storage db is corrupted I guess.

List of Extensions (Sat Apr 24 2021):
Add custom search engine
Amazon.com
Bing
Bing to Google
Bitwarden - Free Password Manager
Bypass Paywalls
Dark Reader
Default Bookmark Folder
Don't touch my tabs! (rel=noopener)
DuckDuckGo
eBay
Firefox Multi-Account Containers
Firefox Private Network
Github1s Extension
GitZip
Google
Hide Files on GitHub
Move Tab Hotkeys
mozlz4-edit
Multithreaded Download Manager
OctoLinker
Octotree - GitHub code tree
Open bookmark in Container Tab menu item
Pushbullet
Reddit Enhancement Suite
Refined GitHub
Simple Translate
SponsorBlock for YouTube - Skip Sponsorships
Stylus
Tab Notes
uBlock Origin
View Image
Violentmonkey BETA
WebNowPlaying Companion - Personal Edit
Wikipedia (en)

**Comment by grahamperrin** [comment]
Thank you, 

### Database

> … Maybe the extension storage db is corrupted …

I'm not a developer but for what it's worth, I don't imagine an issue with SQLite data causing a page context menu item to _never_ appear. 

### Extensions

> … It does show up on a fresh profile, …

I can imagine an **extensions conflict** (I use the phrase loosely). From your opening post <https://github.com/mozilla/multi-account-containers/issues/1996#issue-836955555>:

> … dozens of extensions but this bug exists whether they're disabled or not …

Have you tried the bugged profile with _everything_ disabled except Multi-Account Containers? Specifically: 

1. Disable the extensions
2. quit Firefox
3. start Firefox …

### Environment

I made the page context menu item **not** appear, with a nearly clean profile that had only one or two extensions enabled, by intentionally 'pushing the computer too hard'. This misbehaviour was transient, apparently somewhat random.

**Comment by aminomancer** [comment]
yeah I've tried it without any other extensions enabled. doesn't seem to be that. don't have any other extensions that use the identities API anyway

**Comment by grahamperrin** [comment]
An example of the **page context menu item** _sometimes_ failing to appear. 

A series of screenshots, with the bug evident in the first shot.  

**2021-05-02 02:12:49**

![2021-05-02 02:12:49](https://user-images.githubusercontent.com/192271/116799072-baa0bf00-aaed-11eb-9167-14f6144de7f7.png)

**2021-05-02 02:13:08**

![2021-05-02 02:13:08](https://user-images.githubusercontent.com/192271/116799079-c3919080-aaed-11eb-854b-11fa8cd22668.png)

**2021-05-02 02:13:33**

![2021-05-02 02:13:33](https://user-images.githubusercontent.com/192271/116799086-d2784300-aaed-11eb-8228-f2074450ab12.png)

**2021-05-02 02:14:36**

![2021-05-02 02:14:36](https://user-images.githubusercontent.com/192271/116799097-e0c65f00-aaed-11eb-9dbd-8d524656c617.png)

The fourth shot was used for https://github.com/mozilla/multi-account-containers/issues/796#issuecomment-830720071

* aiming for, but not clicking, a submenu.

**Comment by grahamperrin** [comment]
`about:support` for my _default_ profile: 

[2021-05-03 19:20 about:support – grahamperrin – default profile.txt](https://github.com/mozilla/multi-account-containers/files/6416856/2021-05-03.19.20.about.support.grahamperrin.default.profile.txt)

The list of enabled extensions is long, but please note that I sometimes encounter the bug (missing page context menu item) in other profiles that are minimally extended.

**Comment by grahamperrin** [comment]
Re: https://github.com/mozilla/multi-account-containers/issues/1996#issuecomment-830721081

From <https://github.com/Zegnat/browserext-window-merger/issues/14#issue-873815077>: 

> … two items are missing from the page context menu:
> 
> * Firefox Multi Account Containers
> * Merge Windows

Later I found just one missing (Firefox Multi Account Containers, if I recall correctly). I was unable to take a screenshot, because: 

* after bringing [Spectacle](https://apps.kde.org/spectacle/) to the front, then Firefox to the front (probably when Spectacle hid its window during the countdown to a shot), the menu item became usable. 

For what it's worth, I do get a sense that the bug is less likely to occur after using <kbd>Alt</kbd>-<kbd>Tab</kbd> (from one window to the next).

**Comment by RredCat** [comment]
I also don't have this menu item:

![image](https://user-images.githubusercontent.com/1504503/117122252-607a5500-ad9e-11eb-9308-7c24d97e0048.png)

Below the list of my addons:
List of Extensions (Thu May 06 2021):

Count Tabs
CodeCopy
Free VPN - fastest VPN in the house - disabled
DataScrapper
Mobile View Switcher
1C:Enterprise extension
JetBrains Toolbox Extension
JSONView
Firefox Multi-Account Containers
To Google Translate
ZenMate Free VPN – Best VPN - disabled
Checker Plus for Gmail
Simple Tab Groups
Bitwarden - Free Password Manager
Link Gopher
Copy PlainText
I don't care about cookies
Enhancer for YouTube™
Google
Amazon.com
Bing
DuckDuckGo
eBay
Wikipedia (en)
Udemy Coupons Finder
HTTP Version Indicator
Search by Image
Tab Session Manager
Fess Google Bookmark Extension - disabled
Flagfox
NoScript - disabled
Raindrop.io
Tree Style Tab
Grammarly for Firefox
React Developer Tools
Add-on List-o-matic 9000

**Comment by aminomancer** [comment]
> I also don't have this menu item:

@RredCat If you open a new tab with an actual container (not the default container/non-container) and go to some site, and right click, does it appear for you then? I think the context menu is designed to only appear when the current user context is not 0. In other words, if the selected tab is set to the default user context/container, it will never appear. Though for the last year it hasn't been appearing for me under any circumstances, regardless of the user context, so there may be an issue somewhere.

**Comment by aminomancer** [comment]
By the way, this is totally unrelated, but while testing on a blank profile I noticed the context menu icon could use a bit of work. In the parent process I'm pretty sure extension-loaded SVG icons can use context properties. For some reason, the menu icon style doesn't actually set `-moz-context-properties: fill`, so someone would need to file a bug for Firefox Core/Layout or whatever. Assuming extensions aren't intentionally being denied access to context paint, that is. If that property and `fill: currentColor` were added, then instead of a PNG icon you could use this, and the fill color would automatically adapt to the user's theme:

```
<svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" viewBox="0 0 16 16">
    <g fill="context-fill" fill-opacity="context-fill-opacity">
        <rect width="6" height="6" x="1" y="1" rx="1" />
        <path d="M15 3v2H9V3z" />
        <path d="M11 1h2v6h-2z" />
        <rect width="6" height="6" x="1" y="9" rx="1" />
        <rect width="6" height="6" x="9" y="9" rx="1" />
    </g>
</svg>
```

But I was recently talking to the developer of TreeStyleTabs and he got [around this issue](https://github.com/piroor/treestyletab/commit/702376c693ddbae64e745344762e35ad9e32446b) by using matchMedia() and the theme API to dynamically generate the URL for an icon that conforms to the user's theme's menu text color. So if context properties were out of the question for some reason, that would still be better than having a dark gray icon on a dark gray background.

**Comment by basinilya** [comment]
The trick with page context menu does not work with sites that immediately redirect to other sites. For instance, mail.google.com is set to alway open in container, but in that container I'm already logged in to an account that has Gmail disabled. So I have to logout first, login with the other account and only after that I can uncheck the setting

**Comment by aminomancer** [comment]
>     2. Related to problem 1 above, if the site URL is a redirect URL, it is not possible to access the menu at all. The only solution is to edit the extension data by doing something like opening `about:devtools-toolbox?id=%40testpilot-containers&type=extension` and deleting the site entry.
You should be able to delete any number of site entries by opening the browser action > manage containers > container > manage site list

**Comment by ryanofsky** [comment]
Thanks! I can confirm "browser action > manage containers > container > manage site list" works as another way to remove the container rule without having to actually open the site in the container. I'm not sure how I never noticed the "manage site list" link before.

That leaves the firefox sync issue as the main issue affecting me. And of course it is still frustrating that checking the "remember my decision for this site" checkbox has no effect when you choose to open the site in the current container, as the original issue report describes.

**Comment by aminomancer** [comment]
@ryanofsky Are you sure your containers aren't misaligned between your devices somehow? These are both desktop installations of Firefox, right? Have you tried checking to make sure the containers' values are identical between the installations? I would first try `ContextualIdentityService.getPublicIdentities()` and make sure the `userContextId` values are the same. Then I guess I'd try inspecting the extension as you described and checking that all the stored values are identical between the devices.

**Comment by ryanofsky** [comment]
@aminomancer Yes these are two linux desktop installations in different locations.

The problem happens when I see `siteContainerMap@@_www.google.com` in extension storage, which has value `{"userContextId":"8","neverAsk":false,"identityMacAddonUUID":"356d3da0-9a01-4b4e-9c85-795ccab46120","hostname":"www.google.com"}`

Regardless of whether I delete the key manually in `about:devtools-toolbox?id=%40testpilot-containers&type=extension`, or if I delete it by opening www.google.com in a container and unchecking "Always Open in this Container", or if I delete it by using "manage site list" as I tried most recently, the key always comes back minutes or hours later, and I am prompted to open www.google.com URLs in a container I do not want to use for them. (I am using the temporary containers extension, and I want www.google.com URLs to load in temporary containers like any other URL)

To answer your other questions, I'm not sure how to test the `ContextualIdentityService.getPublicIdentities()` function. I see `ReferenceError: ContextualIdentityService is not defined` pasting it into a debug console. I do know that the "userContextId" value in the `siteContainerMap@@_www.google.com` key is always `8`. I know this because in the past I've tried setting it to nonsense values like 999, which temporarily fixes the problem like deleting the key. But an hour later there will be a sync and the value will be back to 8 and firefox will start prompting to load www.google.com in a container.

**Comment by aminomancer** [comment]
You have to use the console in the [browser toolbox](https://developer.mozilla.org/en-US/docs/Tools/Browser_Toolbox), not the content toolbox. From the content toolbox, go to settings and tick the boxes for enable chrome debugging and enable remote debugging. Then close the content toolbox and open the browser toolbox by hitting ctrl + alt + shift + i. And yeah, the extension debugger's storage inspector only shows you local storage. That is, the contents mapped at `storage.local`, not that which is mapped at `storage.sync`. I don't know much about how this extension works but I guess it probably has its own sync system that hooks up with your FXA account, because storage.sync is severely restricted now both in terms of total extension storage and per-key storage size. That would explain why it still uses local storage, but the local storage seems to revert periodically to some cached or synced state. I would not expect manipulating values from the storage inspector to have any beneficial effect. Instead I would try deleting the container and remaking it. As for temporary containers, that sounds like it could be the source of your problem.

**Comment by ryanofsky** [comment]
@aminomancer Thanks, this is a big help! With the browser toolbox I was able to check `ContextualIdentityService.getPublicIdentities()` on one machine, and I have yet to confirm it on other machine, but I think information there looks ok--it just defining icons, colors, and names for various containers which all look correct, and I think the problem is not due to mismatched `userContextId` values, but instead due to `storage.sync` data periodically overwriting `storage.local` data like you suggested. I was able to check storage data in the `about:devtools-toolbox?id=%40testpilot-containers&type=extension` console with:

```
chrome.storage.sync.get(items => console.log(items))
chrome.storage.local.get(items => console.log(items))
```

And there I see a `siteContainerMap@@_www.google.com` key in `storage.sync` but not `storage.local`. The same key is also listed in the `storage.sync` `deletedSiteList` array. After this, I experimented with assigning the site back to the same container and then unchecking the "Always Open in This Container" menu, and this caused the key to be removed from both sync and local storage, and I'm hopeful this might finally solve the problem.

**Comment by aminomancer** [comment]
That looks promising. I can't use storage.sync for my extensions anymore because of the [storage quotas](https://blog.mozilla.org/addons/2020/07/09/changes-to-storage-sync-in-firefox-79/). I don't think there's a max number of containers or a max number of websites you can bind to a container. Which means it can't all be held in storage.sync since it would eventually run into the 100KB storage.sync quota. I haven't read most of the source code though, so that's just a vague assumption that there is something else going on. I guess it's possible that there's just an unstated limit and maybe it gives you a warning message as you approach it. I don't use containers enough to come anywhere near the limit. But for one of my other extensions, which used to store user notes in storage.sync to sync them between devices, with Fx79 the whole save feature just silently broke down.

**Comment by ryanofsky** [comment]
> Which means it can't all be held in storage.sync since it would eventually run into the 100KB storage.sync quota.

This is interesting to know. In my case `storage.sync` data is only 14KB as pretty-printed JSON, so nowhere near the limit.

> I'm hopeful this might finally solve the problem.

I can confirm a day later after manually re-assigning the same site back to the container on one machine, and then unassigning it with the "Always Open in" checkbox, the syncing problem appears to be fixed for me and the site actually stay unassigned. I do notice the `storage.sync` `deletedSiteList` array is now empty, which make me wonder if the deleted site key could theoretically get resurrected by an older sync on a third machine. But it seems pretty clear now that I hit some syncing edge case which is easily fixed by reassigning and unassigning the site, and there is not some broader sync bug.

On the original issue, I do think it could be alleviated with some UI improvements. The main one would be to fix the partially broken "Remember my decision for this site" checkbox. The checkbox currently only remembers your decision if you click the blue "Open in X container" button, not the gray "Open in container" button. Another useful improvement would be to show the "Always open in This Container" context-menu checkbox on the next page regardless of which button you click, instead of only when you click the blue button.

**Comment by IanVaughan** [comment]
I have the same issue on many sites, here is one example:

* https://app.hibob.com/login/
* macOS Monterey
* FF 104.0.1 (64-bit)

Clicking on "Continue with Google" on this page:
<img width="560" alt="Screenshot 2022-09-09 at 10 31 41" src="https://user-images.githubusercontent.com/158600/189319485-586b55b9-7434-46b5-abc9-874f68113c09.png">

Opens up a new tab with:
<img width="832" alt="Screenshot 2022-09-09 at 10 35 38" src="https://user-images.githubusercontent.com/158600/189320366-c2fb5414-c2a0-4e89-b66f-37018e439472.png">

I would like it to stop doing that and just open in the default container.

**Comment by basinilya** [comment]
Hi @IanVaughan . In the most recent version of the addon it's very easy: 
* click the containers button on the toolbar
* click "Manage containers"
* Click the desired container
* Click "Manage sites list..."
* Hover the mouse above a list entry and click the trash icon

**Comment by IanVaughan** [comment]
Oh thats amazing, so sorry I didn't find that before posting, many thanks!

**Comment by kelvinhammond** [comment]
The fix @basinilya only works for a few seconds for me and then Firefox resets the value. This issue is also synced to all of my browsers. Its annoying that every time I go to google something it asks me if I want to open it in an assigned container.

**Comment by dannycolin** [comment]
@aminomancer did https://github.com/mozilla/multi-account-containers/issues/1996#issuecomment-1242675371 work for you to remove an assignment?

**Comment by aminomancer** [comment]
@dannycolin Yep I'm able to remove site assignments. Feel free to close this. I think the issue @kelvinhammond reported is probably a more general issue with the sync service?

**Comment by dannycolin** [comment]
Closing then. For the sync issue it's most likely related to #1696.
